File: leecode/023_merge-k-sorted-lists.py
'''
合并 k 个排序链表，返回合并后的排序链表。请分析和描述算法的复杂度。

示例:

输入:
[
  1->4->5,
  1->3->4,
  2->6
]
输出: 1->1->2->3->4->4->5->6

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/merge-k-sorted-lists
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
'''
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

import heapq
from heapq import heappop

class Solution:
    def mergeKLists(self, lists) -> ListNode:
        fakeHead = ListNode(0)
        p = fakeHead
        heap =[]
        for node in lists:
            while node:
                heapq.heappush(heap, node.val)
                node = node.next
        
        while len(heap) > 0:
            node = ListNode(heappop(heap))
            p.next = node
            p = p.next

        return fakeHead
if __name__ == "__main__":
    lst1 = ListNode(1)
    lst1.next = ListNode(4)
    lst1.next.next = ListNode(5)

    lst2 = ListNode(1)
    lst2.next = ListNode(3)
    lst2.next.next = ListNode(4)

    lst3 = ListNode(2)
    lst3.next = ListNode(6)

    lst = [lst1, lst2, lst3]

    sol = Solution()
    node = sol.mergeKLists(lst)

    while node.next:
        node = node.next
        print(node.val)

File: leecode/023_merge-k-sorted-lists.java
package suanfa;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;

public class test2 {
	public static void main(String[] args) {
		ListNode ln1 = new ListNode(1);
		ListNode ln2 = new ListNode(4);
		ListNode ln3 = new ListNode(5);
		
		ListNode ln4 = new ListNode(1);
		ListNode ln5 = new ListNode(3);
		ListNode ln6 = new ListNode(4);
		
		ListNode ln7 = new ListNode(2);
		ListNode ln8 = new ListNode(6);
		
		ln1.next = ln2;
		ln2.next = ln3;

		ln4.next = ln5;
		ln5.next = ln6;

		ln7.next = ln8;
		
		ListNode[] lists = new ListNode[] {ln1, ln4, ln7};
		
//		System.out.println("mergeKLists");
//		List<Integer> ret = mergeKLists(lists);
//		for(int item: ret) {
//			System.out.print(item);
//		}
//		System.out.println();
		
		System.out.println("mergeKLists_2");
		ListNode ret2 = mergeKLists_2(lists);
		ListNode node = ret2;
		while(node.next != null) {
			node = node.next;
			System.out.print(node.val);
		}
		
	}
	
	public static List<Integer> mergeKLists(ListNode[] lists) {
	    //利用一个空的链表头方便插入节点。  
	    ListNode fakeHead = new ListNode(0), p = fakeHead;
	    int k = lists.length;
	  
	    
	    PriorityQueue<ListNode> heap = 
	        new PriorityQueue<ListNode>(k, new Comparator<ListNode>() {
	        	public int compare(ListNode a, ListNode b) {
	        		return a.val - b.val;
	        	}
	        });

	    // 从最小堆里将当前最小的节点取出，插入到结果链表中。
	    for (int i = 0; i < k; i++) {
	        if (lists[i] != null) {
	            heap.offer(lists[i]);
	        }
	    }

	    List<Integer> lst = new ArrayList<Integer>();
	    while (!heap.isEmpty()) {
	        ListNode node = heap.poll();
	        lst.add(node.val);
	        
	        p.next = node;
	        p = p.next;
	    
	        // 如果发现该节点后面还有后续节点，将后续节点加入到最小堆里。    
	        if (node.next != null) {
	            heap.offer(node.next);
	        }
	    }
	    
	    return lst;
		  
	}
	
	public static ListNode mergeKLists_2(ListNode[] lists) {
	    //利用一个空的链表头方便插入节点。  
	    ListNode fakeHead = new ListNode(0), p = fakeHead;
	    int k = lists.length;
	  
	    
	    PriorityQueue<ListNode> heap = 
	        new PriorityQueue<ListNode>(k, new Comparator<ListNode>() {
	        	public int compare(ListNode a, ListNode b) {
	        		return a.val - b.val;
	        	}
	        });

	    // 从最小堆里将当前最小的节点取出，插入到结果链表中。
	    for (int i = 0; i < k; i++) {
	        if (lists[i] != null) {
	            heap.offer(lists[i]);
	        }
	    }

	    List<Integer> lst = new ArrayList<Integer>();
	    while (!heap.isEmpty()) {
	        ListNode node = heap.poll();
	        lst.add(node.val);
	        
	        p.next = node;
	        p = p.next;
	    
	        // 如果发现该节点后面还有后续节点，将后续节点加入到最小堆里。    
	        if (node.next != null) {
	            heap.offer(node.next);
	        }
	    }
	    
	    return fakeHead;
	  
	}
}